backend/src/main/java/com/jobwait/persistence/PostgresController.java

```diff
@@ -21,43 +24,57 @@ public class PostgresController extends PersistenceController {
     private String dbPassword = "password";
 
     private Connection getConnection() throws SQLException {
-        return DriverManager.getConnection(jdbcUrl, dbUser, dbPassword);
+        try {
+            return DriverManager.getConnection(jdbcUrl, dbUser, dbPassword);
+        } catch (SQLException e) {
+            throw DatabaseFaults.DatabaseGetConnectionFault();
+        }
     }
 
-    @Override
-    public User getUserFromAuthId(String authId) {
+    private <A> A processAndConvertSQLExceptionToGenericDatabaseFault(SupplierThrowingSQLException<A> func) {
         try {
-            Connection connection = getConnection();
-            PreparedStatement statement = connection.prepareStatement("SELECT * FROM users WHERE authhash = ?");
-            statement.setString(1, authId);
-            ResultSet resultSet = statement.executeQuery();
-            List<User> users = PersistenceUtil.resultSetRowsToAdaptedRows(resultSet, new PostgresUserAdapter());
-            return PersistenceUtil.assertSingleElement(users);
-        } catch (ElementNotFoundException e) {
-            throw new RuntimeException(String.format("Could not find user with authId: %s", authId));
+            return func.get();
         } catch (SQLException e) {
-            throw new RuntimeException(e);
+            throw DatabaseFaults.GenericDatabaseFault();
         }
     }
 
+    @Override
+    public User getUserFromAuthId(String authId) throws FaultException {
+        return processAndConvertSQLExceptionToGenericDatabaseFault(
+                () -> {
+                    try {
+                        Connection connection = getConnection();
+                        PreparedStatement statement = connection
+                                .prepareStatement("SELECT * FROM users WHERE authhash = ?");
+                        statement.setString(1, authId);
+                        ResultSet resultSet = statement.executeQuery();
+                        List<User> users = PersistenceUtil.resultSetRowsToAdaptedRows(resultSet,
+                                new PostgresUserAdapter());
+                        return PersistenceUtil.assertSingleElement(users);
+                    } catch (ElementNotFoundException e) {
+                        throw DatabaseFaults.UserNotFoundFault(authId);
+                    }
+                });
```

Comment:
very gross syntax, I know, see my rant in 0c5f9854e03473cec934f732a1a83786eb215a47. I need a way to delay execution so I can catch SQLExceptions in an abstract way. At best this saves us the mental effort of remembering to put a couple more try/catch blocks which certainly isn't worth it, but I think we should be minimizing try/catch blocks as much as possible (ZIO had this concept of only a single sink at the root of the program, idk if it's applicable to java with its exception throwing)

Reply:
could you clarify how this makes anything clearer? you still catch errors on the way up, what makes this change worthwhile to add, what was the abstraction here?

Reply:
i only catch FaultException on the way up, SqlException is not caught and so it will get thrown. I guess this is a pretty good argument to get rid of SqlException and replace it with a subclass of FaultException. If you really hate this coding style I guessss I can rewrite it, it's fineee, I'll do it

Reply:
to answer your question, the abstraction is me not having to write
```java
try { 
// ...
} catch (SQLException e) {
    throw DatabaseFaults.DatabaseGetConnectionFault();
}
```
in every function of the postgres controller

Reply:
whatever, I've spent too much time on this, repeated code it is

Reply:
ok nice, lgtm, thanks